sdk/nodejs/policy/policy.ts

```diff
@@ -246,6 +246,72 @@ export interface PolicyResource {
     // opts: PolicyResourceOptions;
 }
 
+/**
+ * A helper function that returns a strongly-typed stack validation function.
+ * @param resourceClass A resource class used to filter this check to only resources of the specified class.
+ * @param validate A callback function that validates if a stack violates a policy.
+ */
+export function validateTypedResources<TResource extends Resource>(
+    resourceClass: { new(...rest: any[]): TResource },
+    validate: (
+        resources: q.ResolvedResource<TResource>[],
+        args: StackValidationArgs,
```

Comment:
Right now, the args is being passed as-is, but I am wondering if it'd be better to pass a modified `args` object with `args.resources` filtered to just the resources that match the type filter. The benefit of doing that would be you could then more easily inspect other metadata associated with the strongly-typed resources in `resources` via it's index in the array.

For example, say I'm using this new `validateTypedResources` method with `validateStack` to look at S3 buckets in a strongly-typed way, but also want to inspect those resource's options. If we provided a modified `args` with filtered `args.resources`, then inside the validation function you could do something like:

```typescript
validateStack: validateTypedResources(aws.s3.Bucket, (resources, args, reportViolation) => {
    for (let i = 0; i < resources.length; i++) {
        const bucket = resources[i];
        const bucketOptions = args.resources[i].options;
        ...
    }
}),
```

The downside is then you wouldn't have access to other types of resources from `args.resources` inside that validation function.

Alternatively, we don't pass a modified `args` (continue passing it unmodified) and if you want to look at other metadata associated with a resource, and still inspect the resource in a strongly-typed way, you don't use the `validateTypedResources` helper, but use some of the other helpers like:

```typescript
validateStack: (args, reportViolation) => {
    for (const r of args.resources)
        const bucket = asTypedResource(aws.s3.Bucket, r);
        if (!bucket) continue;
        const bucketOptions = r.options;
        ...
    }
},
```

Reply:
Having more than one way to do something is generally an anti-pattern in software. So if we were to filter the `args.resources`, you'd be forced to use a different approach for writing the policy.

But TBH, I'm personally on-board with having N-different, very-specialized methods for defining policies. (e.g. have `validatedTypedResources` pre-filter `args.resources` to _just_ those that match the provided resource type.

The reason being is that all of these helper methods are designed to make it writing policies more expressive, and more declarative. A policy that can be expressed in 10 lines of code is more useful than the same functionality expressed in 20 lines of code, assuming there is no sacrifice in readability/clarity.

The only pushback I have though is that if we do have multiple ways to do something -- e.g. `validateTypedResources` and other similar helper methods -- we need to make them easier to use/mix and match. The name `validateTypedResources` doesn't really give an indication of what specifically is going on. But perhaps a different name like:

`validateStack: validateStackResourcesWithType(aws.s3.Bucket, (resources, args, reportViolation) => {`
or 
`validateStack: onlyResourcesWithType(aws.s3.Bucket, (resources, args, reportViolation) => {`
 etc.

Make sense?

Reply:
I agree the naming could be more specific. I dont have a problem with the multiple helpers at least for now. I think we need to see how people are writing policies and what makes the most sense before making a call on whats the "best helper"